compiler/resolution/fixupExports.cpp

```diff
@@ -0,0 +1,425 @@
+
+#include "astutil.h"
+#include "baseAST.h"
+#include "passes.h"
+#include "resolution.h"
+#include "resolveFunction.h"
+#include "scopeResolve.h"
+#include "stmt.h"
+#include "stringutil.h"
+#include "symbol.h"
+
+#include <map>
+#include <vector>
+
+static std::map<FnSymbol*, FnSymbol*> wrapperMap;
+static std::map<const char*, FnSymbol*> conversionCallMap;
+
+static void attemptFixups(FnSymbol* fn);
+static bool needsFixup(ArgSymbol* as);
+static bool needsFixup(Type* t);
+static void validateFormalIntent(FnSymbol* fn, int idx);
+static void validateReturnIntent(FnSymbol* fn);
+static FnSymbol* createWrapper(FnSymbol* fn);
+static Type* getTypeForFixup(Type* t, bool ret=false);
+static const char* getConversionCallName(Type* srctype, Type* dsttype);
+static FnSymbol* getConversionCall(Type* srctype, Type* dsttype);
+static FnSymbol* resolveConversionCall(Type* srctype, Type* dsttype);
+static VarSymbol* fixupFormal(FnSymbol* wrapper, int idx);
+static void changeRetType(FnSymbol* wrapper);
+static void insertUnwrappedCall(FnSymbol* wrapper, FnSymbol* fn,
+                                const std::vector<VarSymbol*>& tmps);
+
+//
+// Some type information is lost during this pass (IE, converting `string`
+// formals to `c_string`, and this information might be useful when
+// generating code in later passes.
+//
+FnSymbol* getUnwrappedFunction(FnSymbol* wrapper) {
+  std::map<FnSymbol*, FnSymbol*>& m = wrapperMap;
+  if (m.find(wrapper) == m.end()) {
+    return NULL;
+  }
+  return m[wrapper];
+}
+
+void fixupExportedFunctions(const std::vector<FnSymbol*>& fns) {
+  for (size_t i = 0; i < fns.size(); i++) {
+    FnSymbol* fn = fns[i];
+    SET_LINENO(fn);
+    attemptFixups(fn);
+  }
+}
+
+void fixupExportedFunction(FnSymbol* fn) {
+  attemptFixups(fn);
+}
+
+Type* getCharPtrType(void) {
+  static Type* result = NULL;
+
+  if (result != NULL) { return result; }
+
+  const char* modname = "ExternalString";
+  const char* symname = "chpl__exportTypeCharPtr";
+  ModuleSymbol* esm = NULL;
+  Symbol* alias = NULL;
+
+  // Get a handle to the ExternalString module.
+  forv_Vec(ModuleSymbol, md, allModules) {
+    if (md->modTag == MOD_INTERNAL && !strcmp(md->name, modname)) {
+      esm = md;
+      break;
+    }
+  }
+
+  INT_ASSERT(esm != NULL);
+
+  // Get a handle to the c_ptr(c_char) type from our global alias.
+  for_alist(expr, esm->block->body) {
+    DefExpr* de = toDefExpr(expr);
+    if (de) {
+      Symbol* sym = de->sym;
+      if (!strcmp(sym->name, symname)) {
+        alias = sym;
+        break;
+      }
+    }
+  }
+
+  SymExpr* rctx = new SymExpr(alias);
+
+  // Insert into tree and resolve, just in case.
+  chpl_gen_main->insertAtHead(rctx);
+  resolveExpr(rctx);
+  rctx->remove();
+
+  INT_ASSERT(alias->type != NULL);
+  result = alias->type;
+
+  return result;
+}
+
+static void attemptFixups(FnSymbol* fn) {
+  std::vector<VarSymbol*> tmps;
+  FnSymbol* wrapper = NULL;
+
+  if (not fn->hasFlag(FLAG_EXPORT)) { return; }
```

Comment:
As far as I know we don't typically use the alternative operators like ``not``, so for consistency I'd recommend just using ``!`` for now.